TShockAPI/Commands.cs

```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+				if (ban.ExpirationDateTime < DateTime.UtcNow)
+				{
+					args.Player.SendMessage($"{"Banned expired:".Color(Utils.BoldHighlight)} {ban.ExpirationDateTime.ToString("yyyy/MM/dd").Color(Utils.RedHighlight)} ({ban.GetPrettyExpirationString().Color(Utils.YellowHighlight)} ago)", Color.White);
+				}
+				else
+				{
+					string remaining;
+					if (ban.ExpirationDateTime == DateTime.MaxValue)
+					{
+						remaining = "Never".Color(Utils.YellowHighlight);
+					}
+					else
+					{
+						remaining = $"{ban.GetPrettyExpirationString().Color(Utils.YellowHighlight)} remaining";
+					}
 
-							if (!parsedOkay)
-							{
-								args.Player.SendErrorMessage("Invalid time format. Example: 10d 5h 3m 2s.");
-								args.Player.SendErrorMessage("Use 0 (zero) for a permanent ban.");
-								return;
-							}
-						}
+					args.Player.SendMessage($"{"Banned expires:".Color(Utils.BoldHighlight)} {ban.ExpirationDateTime.ToString("yyyy/MM/dd").Color(Utils.RedHighlight)} ({remaining})", Color.White);
+				}
+			}
 
-						// If a reason exists, use the given reason.
-						if (args.Parameters.Count > 3)
-						{
-							banReason = String.Join(" ", args.Parameters.Skip(3));
-						}
+			void AddBan()
+			{
+				if (!args.Parameters.TryGetValue(1, out string target))
+				{
+					args.Player.SendMessage($"Invalid Ban Add syntax. Refer to {"ban help add".Color(Utils.BoldHighlight)} for details on how to use the {"ban add".Color(Utils.BoldHighlight)} command", Color.White);
+					return;
+				}
 
-						// Good case: Online ban for matching character.
-						if (players.Count == 1)
-						{
-							TSPlayer target = players[0];
+				bool exactTarget = args.Parameters.Any(p => p == "-e");
+				bool banAccount = args.Parameters.Any(p => p == "-a");
+				bool banUuid = args.Parameters.Any(p => p == "-u");
+				bool banName = args.Parameters.Any(p => p == "-n");
+				bool banIp = args.Parameters.Any(p => p == "-ip");
 
-							if (target.HasPermission(Permissions.immunetoban) && !callerIsServerConsole)
-							{
-								args.Player.SendErrorMessage("Permission denied. Target {0} is immune to ban.", target.Name);
-								return;
-							}
+				string reason = "Banned.";
+				string duration = null;
+				DateTime expiration = DateTime.MaxValue;
+				AddBanResult banResult;
 
-							targetGeneralizedName = target.Name;
-							success = TShock.Bans.AddBan(target.IP, target.Name, target.UUID, target.Account?.Name ?? "", banReason, false, args.Player.Account.Name,
-								banLengthInSeconds == 0 ? "" : DateTime.UtcNow.AddSeconds(banLengthInSeconds).ToString("s"));
+				//This is hacky, but because the flag values can go at any parameter, this is forcing the ordering of 'reason' and 'duration'
```

Comment:
So the purpose here is to allow the flags to go in any order, and then make duration and reason order dependent? If so, you should like, reword to something that emphasizes the _goal_.

```suggestion
				//This is hacky, because we want the flag values to be at any position, but this forces us to order reason and duration consecutively
```